File: src/parsers/hash.ts
/**
 * SHA-256 project name hashing for privacy protection.
 *
 * This module provides stable, short hashes of project paths and session IDs
 * to protect user privacy in shared/corporate environments.
 */

import { createHash } from "node:crypto";

/**
 * Creates a stable, short hash of project paths for privacy.
 *
 * Uses SHA-256 to create an 8-character hexadecimal hash that is:
 * - Stable: Same input always produces the same output
 * - Collision-resistant: ~4 billion possible values (16^8)
 * - Privacy-preserving: Original paths cannot be reverse-engineered
 * - Short: 8 characters instead of full paths like "/home/user/project/very-long-name"
 *
 * @param projectPath - Full project path or session ID to hash
 * @param enabled - Whether hashing is enabled (default: true)
 * @returns 8-character hexadecimal hash, or original path if hashing disabled
 *
 * @example
 * ```ts
 * // With hashing enabled (default)
 * hashProjectName("/home/user/project/my-app") // "a3f7b2c1"
 * hashProjectName("/home/user/project/my-app") // "a3f7b2c1" (stable!)
 *
 * // With hashing disabled
 * hashProjectName("/home/user/project/my-app", false) // "/home/user/project/my-app"
 *
 * // Session IDs work too
 * hashProjectName("/home/user/project/my-app/session-abc123") // "d8e4f6a2"
 * ```
 */
export async function hashProjectName(
  projectPath: string,
  enabled: boolean = true
): Promise<string> {
  // Return original path if hashing is disabled
  if (!enabled) {
    return projectPath;
  }

  // Encode the string to bytes
  const encoder = new TextEncoder();
  const data = encoder.encode(projectPath);

  // Use Web Crypto API for SHA-256 hashing
  const hashBuffer = await crypto.subtle.digest("SHA-256", data);

  // Convert buffer to hex string and take first 8 characters
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  const hashHex = hashArray
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");

  return hashHex.substring(0, 8);
}

/**
 * Synchronous version of hashProjectName for contexts where async is not available.
 *
 * Note: This uses Node.js crypto module for synchronous hashing. In most cases,
 * prefer using the async version (hashProjectName) which uses Web Crypto API.
 *
 * @param projectPath - Full project path or session ID to hash
 * @param enabled - Whether hashing is enabled (default: true)
 * @returns 8-character hexadecimal hash, or original path if hashing disabled
 */
export function hashProjectNameSync(
  projectPath: string,
  enabled: boolean = true
): string {
  // Return original path if hashing is disabled
  if (!enabled) {
    return projectPath;
  }

  // Use Node.js crypto module for sync hashing
  const hash = createHash("sha256");
  hash.update(projectPath);
  const digest = hash.digest("hex");

  return digest.substring(0, 8);
}

/**
 * Batch hash multiple project paths efficiently.
 *
 * @param projectPaths - Array of project paths to hash
 * @param enabled - Whether hashing is enabled (default: true)
 * @returns Array of hashed project paths in the same order as input
 */
export async function hashProjectNames(
  projectPaths: string[],
  enabled: boolean = true
): Promise<string[]> {
  if (!enabled) {
    return projectPaths;
  }

  return Promise.all(
    projectPaths.map((path) => hashProjectName(path, enabled))
  );
}

/**
 * Batch hash multiple project paths synchronously.
 *
 * @param projectPaths - Array of project paths to hash
 * @param enabled - Whether hashing is enabled (default: true)
 * @returns Array of hashed project paths in the same order as input
 */
export function hashProjectNamesSync(
  projectPaths: string[],
  enabled: boolean = true
): string[] {
  if (!enabled) {
    return projectPaths;
  }

  return projectPaths.map((path) => hashProjectNameSync(path, enabled));
}

/**
 * Validate if a string looks like a hashed project path.
 *
 * Hashed project paths should be exactly 8 hexadecimal characters.
 *
 * @param value - Value to validate
 * @returns True if the value looks like a hashed project path
 */
export function isHashedProjectPath(value: string): boolean {
  return /^[a-f0-9]{8}$/.test(value);
}

/**
 * Convert a hash back to a placeholder identifier.
 *
 * This does NOT reverse the hash (SHA-256 is one-way), but provides
 * a consistent placeholder format for display purposes.
 *
 * @param hash - The hashed project path
 * @returns A formatted placeholder string
 */
export function formatHashedProjectPath(hash: string): string {
  if (!isHashedProjectPath(hash)) {
    return hash;
  }
  return `<project:${hash}>`;
}

File: apps/cli/src/util/hash.rs
/**
 * SHA-256 project-name hashing for privacy protection.
 *
 * Uses the `sha2` crate to produce an 8-character lowercase hex hash
 * identical to the TS implementation:
 * `crypto.createHash('sha256').update(path).digest('hex').slice(0, 8)`
 */

use sha2::{Digest, Sha256};

/// Return the first 8 lowercase hex chars of `SHA-256(input)`.
pub fn hash_project_name_sync(project_path: &str, enabled: bool) -> String {
    if !enabled {
        return project_path.to_string();
    }
    let mut hasher = Sha256::new();
    hasher.update(project_path.as_bytes());
    let result = hasher.finalize();
    // Take first 4 bytes → 8 hex chars
    format!("{:02x}{:02x}{:02x}{:02x}", result[0], result[1], result[2], result[3])
}

/// Batch hash multiple paths synchronously.
pub fn hash_project_names_sync(paths: &[&str], enabled: bool) -> Vec<String> {
    if !enabled {
        return paths.iter().map(|p| p.to_string()).collect();
    }
    paths.iter().map(|p| hash_project_name_sync(p, true)).collect()
}

/// Dedup key: first 16 hex chars of SHA-256 of the joined raw key.
pub fn make_dedup_key(raw_key: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(raw_key.as_bytes());
    let result = hasher.finalize();
    // Take first 8 bytes → 16 hex chars
    format!(
        "{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
        result[0], result[1], result[2], result[3], result[4], result[5], result[6], result[7]
    )
}

/// Check if a string looks like a hashed project path (8 lowercase hex chars).
pub fn is_hashed_project_path(value: &str) -> bool {
    value.len() == 8
        && value.chars().all(|c| c.is_ascii_hexdigit())
        && value.chars().filter(|c| c.is_ascii_alphabetic()).all(|c| c.is_ascii_lowercase())
}

/// Format a hash as a placeholder, or return original if not a valid hash.
pub fn format_hashed_project_path(hash: &str) -> String {
    if is_hashed_project_path(hash) {
        format!("<project:{}>", hash)
    } else {
        hash.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_ts_hash_for_known_inputs() {
        // Verified with TS: createHash('sha256').update('/home/user/project/test').digest('hex').slice(0,8)
        assert_eq!(hash_project_name_sync("/home/user/project/test", true), "b49e9761");
        assert_eq!(hash_project_name_sync("/home/user/project/my-app", true), "9b339b46");
    }

    #[test]
    fn disabled_returns_original() {
        assert_eq!(hash_project_name_sync("/home/user/project/my-app", false), "/home/user/project/my-app");
    }

    #[test]
    fn stable_across_calls() {
        let h1 = hash_project_name_sync("/home/user/project/my-app", true);
        let h2 = hash_project_name_sync("/home/user/project/my-app", true);
        assert_eq!(h1, h2);
        assert_eq!(h1.len(), 8);
    }

    #[test]
    fn different_inputs_different_hashes() {
        assert_ne!(
            hash_project_name_sync("/a", true),
            hash_project_name_sync("/b", true),
        );
    }

    #[test]
    fn lowercase_hex_only() {
        let h = hash_project_name_sync("/some/path", true);
        assert!(h.chars().all(|c| c.is_ascii_hexdigit()));
        assert!(h.chars().filter(|c| c.is_ascii_alphabetic()).all(|c| c.is_ascii_lowercase()));
    }

    #[test]
    fn empty_string_works() {
        let h = hash_project_name_sync("", true);
        assert_eq!(h.len(), 8);
    }

    #[test]
    fn batch_matches_individual() {
        let paths = ["/a", "/b", "/c"];
        let batch = hash_project_names_sync(&paths, true);
        let individual: Vec<String> = paths.iter().map(|p| hash_project_name_sync(p, true)).collect();
        assert_eq!(batch, individual);
    }

    #[test]
    fn is_hashed_project_path_valid() {
        assert!(is_hashed_project_path("a3f7b2c1"));
        assert!(is_hashed_project_path("00000000"));
        assert!(is_hashed_project_path("abcdef01"));
        assert!(!is_hashed_project_path("A3F7B2C1")); // uppercase
        assert!(!is_hashed_project_path("a3f7b2c"));   // too short
        assert!(!is_hashed_project_path("a3f7b2c11")); // too long
    }

    #[test]
    fn format_hashed_project_path_works() {
        assert_eq!(format_hashed_project_path("a3f7b2c1"), "<project:a3f7b2c1>");
        assert_eq!(format_hashed_project_path("/home/user"), "/home/user");
    }

    #[test]
    fn dedup_key_16_chars() {
        let key = hash_project_name_sync("/home/user/project/my-app", true);
        let dk = make_dedup_key(&format!("ccusage|{}|daily|2025-01-05|{}|2025-01-05", "test-machine", key));
        assert_eq!(dk.len(), 16);
    }
}
